### PR TITLE
Added member role and reworked roles update functions

### PR DESCRIPTION
**Pull request by ioaNikas**
Implement front roles for prod
- [x] Remove ethereum operations from organization roles
- [x] SuperAdmin can do everything, Admin can invite/decline users, Member can read documents
- [x] Added new role with less privileges: member

Fix issue #1097

### CODE REVIEW

File: libs/organization/src/lib/+state/organization.query.ts [hunk 1]
```diff
@@ -36,22 +32,36 @@ export class OrganizationQuery extends Query<OrganizationState> {
     protected store: OrganizationStore,
     private permissionsQuery: PermissionsQuery,
     protected db: FireQuery
```
Comment (GrandSchtroumpf) [inline libs/organization/src/lib/+state/organization.query.ts]: Remove dependency to FireQuery

File: libs/organization/src/lib/+state/organization.query.ts [hunk 2]
```diff
@@ -36,22 +32,36 @@ export class OrganizationQuery extends Query<OrganizationState> {
...
-  ]).pipe(
-    map(([members, superAdmins]) => {
-      return members.map(member => ({
-        ...member,
-        role: superAdmins.includes(member.uid) ? UserRole.admin : UserRole.member
-      }));
+  public membersWithRole$: Observable<OrganizationMember[]> = this.members$.pipe(
+    map(orgMembers => {
+      return orgMembers.map(orgMember => {
+        const permissions = this.permissionsQuery.getValue();
```
Comment (GrandSchtroumpf) [inline libs/organization/src/lib/+state/organization.query.ts]: What if permission change ? Why to you request the value for each member ?
```typescript
combineLatest([
  this.members$, // when OrganizationQuery is EntityQuery
  this.permissionQuery.selectAll()
]).pipe(
  map(([ members, permissions ]) => membres.map(member => addRole(member, permissions)))
)
```

Add role is a pure function somewhere on the top of the file : 
```typescript
function addRole(member: Member, permissions: Permissions) {
  if (permissions.superAdmins.includes(member.id)) {
    return { ...member, role: ... }
  } else if (...) { .. }
}
```

File: libs/organization/src/lib/components/member-form-role/member-form-role.component.html [hunk 3]
```diff
@@ -2,22 +2,26 @@ <h2>{{name}}</h2>
...
     <mat-label>Role</mat-label>
-    <mat-select [formControl]="role" (valueChange)="changeRole($event)">
+    <mat-select [formControl]="role">
       <mat-option value="member"> Member </mat-option>
       <mat-option value="admin"> Admin </mat-option>
+      <mat-option value="superAdmin"> Super Admin </mat-option>
     </mat-select>
   </mat-form-field>
+  <ng-container *ngIf="displayWarningMessage">
+    <div><mat-icon>error_outline</mat-icon></div>
```
Comment (GrandSchtroumpf) [inline libs/organization/src/lib/components/member-form-role/member-form-role.component.html]: Remove div

File: libs/organization/src/lib/components/member-form-role/member-form-role.component.ts [hunk 4]
```diff
@@ -19,79 +20,45 @@ export class MemberFormRoleComponent {
...
   ) {}
 
   get control() {
     return this.controlContainer.control;
   }
 
   public get name() {
-    const {name} = this.control.value;
+    const { name } = this.control.value;
     return name;
```
Comment (GrandSchtroumpf) [inline libs/organization/src/lib/components/member-form-role/member-form-role.component.ts]: `return this.control.value.name` ^^

File: libs/organization/src/lib/components/member-form-role/member-form-role.component.ts [hunk 5]
```diff
@@ -19,79 +20,45 @@ export class MemberFormRoleComponent {
...
   }
 
   public get role() {
     return this.control.get('role');
   }
 
+  public get userId() {
+    const { uid } = this.control.value;
```
Comment (GrandSchtroumpf) [inline libs/organization/src/lib/components/member-form-role/member-form-role.component.ts]: `return this.control.value.uid` ^^

File: libs/organization/src/lib/components/member-form-role/member-form-role.component.ts [hunk 6]
```diff
@@ -19,79 +20,45 @@ export class MemberFormRoleComponent {
...
-        redirectName: 'Back to Administration',
-        redirectRoute: `/layout/o/organization/${orgId}/members`,
-      }
-    }
-
-    this.walletService.setTx(tx);
-    this.walletService.setTxFeedback(feedback);
-    this.router.navigateByUrl('/layout/o/account/wallet/send');
+  /** Display a message if a Super Admin is about to downgrade himself. */
+  public get displayWarningMessage() {
```
Comment (GrandSchtroumpf) [inline libs/organization/src/lib/components/member-form-role/member-form-role.component.ts]: `willDowngrade` ?

File: libs/organization/src/lib/components/organization-signer-repertory/organization-signer-repertory.component.ts [hunk 7]
```diff
@@ -68,8 +68,20 @@ export class OrganizationSignerRepertoryComponent {
 
     const operationMembers: OperationMember[] = this._members.map(member => {
       const operationIds = getOperationIds(member);
-      const isAdmin = this.permissionQuery.isUserSuperAdmin(member.uid);
-      const role = isAdmin ? UserRole.admin : UserRole.member;
+      let role: UserRole;
```
Comment (GrandSchtroumpf) [inline libs/organization/src/lib/components/organization-signer-repertory/organization-signer-repertory.component.ts]: DRY : You have already written something like that. Create reusable functions.

File: libs/organization/src/lib/permissions/+state/permissions.query.ts [hunk 8]
```diff
@@ -9,31 +9,37 @@ import { map } from 'rxjs/operators';
...
   }
 
-  public get superAdminCount() {
-    return this.getValue().superAdmins.length;
-  }
-
-  /** Checks if the connected user is either SuperAdmin or Admin of his organization */
-  public get isOrgAdmin$(): Observable<boolean> {
+  /** Checks if the connected user is admin of his organization. */
+  public get isAdmin$(): Observable<boolean> {
```
Comment (GrandSchtroumpf) [inline libs/organization/src/lib/permissions/+state/permissions.query.ts]: Observable should not be getters. Create a property instead.

File: libs/organization/src/lib/permissions/+state/permissions.query.ts [hunk 9]
```diff
@@ -9,31 +9,37 @@ import { map } from 'rxjs/operators';
...
-  /** Return an array containing the userIds of the superAdmins of the organization */
-  public superAdmins$: Observable<string[]> = this.select(state => state.superAdmins);
 
   constructor(protected store: PermissionsStore, private auth: AuthQuery) {
     super(store);
   }
 
-  /** Checks if the connected user is SuperAdmin of his organization */
+  /** Checks if the connected user is superAdmin of his organization. */
   public get isSuperAdmin$(): Observable<boolean> {
```
Comment (GrandSchtroumpf) [inline libs/organization/src/lib/permissions/+state/permissions.query.ts]: Observable should not be getters. Create a property instead.

File: libs/organization/src/lib/permissions/+state/permissions.query.ts [hunk 10]
```diff
@@ -9,31 +9,37 @@ import { map } from 'rxjs/operators';
...
     return this.isSuperAdmin$.pipe(
       map(isSuperAdmin => isSuperAdmin || this.getValue().admins.includes(this.auth.userId))
     );
   }
 
-  /** Checks if the user is SuperAdmin of his organization */
-  public isUserSuperAdmin(userId: string) {
-    return this.getValue().superAdmins.includes(userId);
+  /** Checks if the connected user is either member of his organization. */
+  public get isOrgMember$(): Observable<boolean> {
```
Comment (GrandSchtroumpf) [inline libs/organization/src/lib/permissions/+state/permissions.query.ts]: Observable should not be getters. Create a property instead.

File: libs/organization/src/lib/permissions/+state/permissions.service.ts [hunk 11]
```diff
@@ -37,34 +41,39 @@ export class PermissionsService {
...
   /** Update roles of members of the organization */
-  public async updateMembersRole(members: OrganizationMember[]) {
+  public async updateMembersRole(organizationMembers: OrganizationMember[]) {
     const orgId = this.query.getValue().orgId;
     const orgPermissionsDocRef = this.db.doc<Permissions>(`permissions/${orgId}`).ref;
 
     return this.db.firestore.runTransaction(async tx => {
       const orgPermissionsDoc = await tx.get(orgPermissionsDocRef);
-      let {superAdmins, admins} = orgPermissionsDoc.data() as Permissions;
+      let { superAdmins, admins, members } = orgPermissionsDoc.data() as Permissions;
```
Comment (GrandSchtroumpf) [inline libs/organization/src/lib/permissions/+state/permissions.service.ts]: Why using `let` ?

File: libs/organization/src/lib/permissions/+state/permissions.service.ts [hunk 12]
```diff
@@ -37,34 +41,39 @@ export class PermissionsService {
...
+          superAdmins = [...superAdmins, uid];
         }
-        // Case of role = 'member': we remove the memberId of superAdmins and insert it in admins
-        else {
-          if (!isAdmin(uid) && isSuperAdmin(uid)) {
-            superAdmins = superAdmins.filter(superAdmin => superAdmin !== uid);
-            admins = [...admins, uid]
-          };
+        // Case of role = 'admin': we remove the member id from superAdmins/members and insert it in admins
+        if (role === UserRole.admin && !isAdmin(uid)) {
```
Comment (GrandSchtroumpf) [inline libs/organization/src/lib/permissions/+state/permissions.service.ts]: Looks super difficult to maintain this model.

File: libs/organization/src/lib/+state/organization.query.ts (hunk 2 above)
Comment (ioaNikas) [reply libs/organization/src/lib/+state/organization.query.ts]: So much easier to understand with this, thanks a lot !

File: libs/organization/src/lib/components/organization-signer-repertory/organization-signer-repertory.component.ts (hunk 7 above)
Comment (ioaNikas) [reply libs/organization/src/lib/components/organization-signer-repertory/organization-signer-repertory.component.ts]: As we only need to have members with role here, I'm passing `membersWithRole$` from member.query to this component so we just have to add the `operationIds` property in the member \o/.

File: apps/backend-functions/src/data/internals.ts [hunk 13]
```diff
@@ -85,15 +53,21 @@ export function getCount(collection: string): Promise<number> {
...
+  const permissions = await getDocument<PermissionsDocument>(`permissions/${organizationId}`);
 
-  if (!permissionsDoc.exists) {
+  if (!permissions) {
     throw new Error(`organization: ${organizationId} does not exists`);
   }
 
-  const { superAdmins } = permissionsDoc.data() as OrganizationPermissions;
-  return superAdmins;
+  console.log('Permissions', {permissions});
```
Comment (GrandSchtroumpf) [inline apps/backend-functions/src/data/internals.ts]: Is it supposed to stay ?

File: apps/backend-functions/src/invitation.ts [hunk 14]
```diff
@@ -235,18 +239,29 @@ async function onInvitationFromUserToJoinOrgCreate({
...
-  const superAdminIds = await getSuperAdminIds(organization.id);
+  const adminIds = await getAdminIds(organization.id);
 
-  const superAdmins = await Promise.all(superAdminIds.map(getUser));
+  const admins = await Promise.all(adminIds.map(getUser));
   // const validSuperAdminMails = superAdminsMails.filter(adminEmail => !!adminEmail);
 
   // send invitation pending email to user
-  await sendMailFromTemplate(userJoinOrgPendingRequest(userData.email, organization.name, userData.name!));
+  await sendMailFromTemplate(
```
Comment (GrandSchtroumpf) [inline apps/backend-functions/src/invitation.ts]: create a variable : 
```typescript
const request = userJoinOrgPendingRequest(userData.email, organization.name, userData.name!)
rawait sendMailFromTemplate(request)
```

File: apps/catalog/marketplace/marketplace/src/app/layout/context-menu.ts [hunk 15]
```diff
@@ -103,6 +104,42 @@ export const CONTEXT_MENU_AFM = [ // TODO #1146
       }
     ]
   },
+  {
+    route: '/layout/o/organization/:orgId/edit',
```
Comment (GrandSchtroumpf) [inline apps/catalog/marketplace/marketplace/src/app/layout/context-menu.ts]: Use `appRoute`: `${appsRoute}/organization/:orgId/edit`

File: libs/organization/src/lib/+state/organization.service.ts [hunk 16]
```diff
@@ -135,7 +136,7 @@ export class OrganizationService extends CollectionService<OrganizationState> {
   async onCreate(org: Organization, { write }: WriteOptions) {
     const user = this.authQuery.user;
     const orgId: string = org.id;
-    const permissions = createPermissions({ orgId, superAdmins: [user.uid] });
+    const permissions = createOrgPermissions({ orgId, roles: { [user.uid]: UserRole.superAdmin } });
```
Comment (GrandSchtroumpf) [inline libs/organization/src/lib/+state/organization.service.ts]: Maybe in two lines to make things more readable...

File: libs/organization/src/lib/components/member-form-role/member-form-role.component.html [hunk 17]
```diff
@@ -2,22 +2,26 @@ <h2>{{name}}</h2>
...
+    (click)="destroyWallet()">
     <mat-icon>delete</mat-icon>
     Destroy Wallet
   </button>
 </ng-container>
-<ng-template #cannotChangeRole>
-  <div><mat-icon>error_outline</mat-icon></div>
-  <p>You can't change the role of the last Admin because an organization always need at least one Admin.</p>
-  <p>If you want to change the role of this Admin, please promote another Admin before.</p>
+<ng-template #cannotChangeRole class="de">
```
Comment (GrandSchtroumpf) [inline libs/organization/src/lib/components/member-form-role/member-form-role.component.html]: What is class "de" ??? German class ?
Also I'm pretty sure, class doesn't have effect on ng-template...

Comment (ioaNikas) [reply libs/organization/src/lib/components/member-form-role/member-form-role.component.html]: Sorry it's a very dirty remnant that I forget to flush.

**Pull request merged**